Fix matchstick prediction issues

beck-thompson opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
Matchstick was always miss predicting getting inserted into a container. There was an outdated method that I fixed as well.

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
The only sussy thing I had to do was always make the event handled if a matchstick is used on a matchbox. If you don't do that and check for it being lit and then making it list if its not, it starts misrepresenting on the second round because the match becomes lit and the event wouldn't be handled.

![image](https://github.com/user-attachments/assets/cf811a0c-f273-4a9d-8311-261439fd76c1)
\* This works the first time but the second time it will never trigger so it won't get handled and gets inserted.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.
-->

https://github.com/user-attachments/assets/a8bdfa3f-ea6f-44f3-b894-db2dc7fb6bbb

## Requirements
<!-- 
Due to influx of PR's we require to ensure that PR's are following the correct guidelines.

Please take a moment to read these if its your first time.

Check the boxes below to confirm that you have in fact seen these (put an X in the brackets, like [X]):
-->
- [X] I have read and I am following the [Pull Request Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html). I understand that not doing so may get my pr closed at maintainer’s discretion
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

Breaking changes:
The match / matchbox systems and components have been moved from `Content.xxx.Light` to `Content.xxx.IgnitionSource`. `TryIgnite`'s user parameter is now optional and it will check if the match can be lit before attempting to light it. Matches also now use `IgnitionSource`es to light things on fire instead of their own home grown system

Winkarst-cpu commented:
Oh, I think you need to update yml of matches too...

beck-thompson commented:
Your right but I have no idea why its failing... Isn't the default variable name lowercase? It was:
`[DataField("duration")]` but the name was still just `Duration` which should default to the same thing?

Winkarst-cpu commented:
Try removing read-only access, I don't know.

Winkarst-cpu commented:
Uh... The old one was read-only too...

beck-thompson commented:
OK we couldn't figure out why it always fails when the datafield names are removed (Even when the names would still work...). Just leaving them there!

metalgearsloth commented on Content.Server/Light/EntitySystems/MatchstickSystem.cs:
```diff
@@ -1,123 +1,47 @@
+            var xform = Transform(uid);
 
-            if (!isHotEvent.IsHot)
+            if (xform.GridUid is not { } gridUid)
                 return;
```
Return means this will terminate the entire method if any single match is not on a grid.

metalgearsloth commented on Content.Shared/Lightning/Components/MatchboxComponent.cs:
```diff
@@ -0,0 +1,9 @@
+namespace Content.Shared.Light.Components;
+
+// TODO make changes in icons when different threshold reached
+// e.g. different icons for 10% 50% 100%
+[RegisterComponent]
+public sealed partial class MatchboxComponent : Component
+{
+}
```
Needs networking.

metalgearsloth commented on Content.Shared/Lightning/Components/MatchstickComponent.cs:
```diff
@@ -0,0 +1,35 @@
+    /// </summary>
+    [DataField("duration")]
+    public int Duration = 10;
+
+    /// <summary>
+    ///     The time that the match will burn out. If null, that means the match is unlit.
+    /// </summary>
+    [AutoNetworkedField]
+    public TimeSpan? TimeMatchWillBurnOut = null;
+
```
Needs autopausedfield.

metalgearsloth commented on Content.Shared/Lightning/EntitySystems/MatchboxSystem.cs:
```diff
@@ -0,0 +1,29 @@
+using Content.Shared.Storage.EntitySystems;
+using Content.Shared.Interaction;
+using Content.Shared.Smoking;
+
+namespace Content.Shared.Light.EntitySystems
+{
```
file-scoped namespaces.

slarticodefast commented on Content.Server/Light/EntitySystems/MatchstickSystem.cs:
```diff
@@ -1,123 +1,42 @@
-using Content.Shared.Item;
 using Content.Shared.Smoking;
-using Content.Shared.Temperature;
+using Content.Shared.Light.Components;
+using Content.Server.Atmos.EntitySystems;
+using Content.Shared.Light.EntitySystems;
```
not in alphabetic order 😢

slarticodefast commented on Content.Server/Light/EntitySystems/MatchstickSystem.cs:
```diff
@@ -1,123 +1,42 @@
 
-            if (EntityManager.TryGetComponent(uid, out AppearanceComponent? appearance))
+            // Check if the match has expired.
+            var burnoutTime = match.TimeMatchWillBurnOut;
```
why assign this as a variable if you only use it once?

slarticodefast commented on Content.Server/Light/EntitySystems/MatchstickSystem.cs:
```diff
@@ -1,123 +1,42 @@
             {
-                _appearance.SetData(uid, SmokingVisuals.Smoking, component.CurrentState, appearance);
+                SetState(uid, match, SmokableState.Burnt);
+                match.TimeMatchWillBurnOut = null;
```
do we have to set it to null?
shouldn't the state check at the beginning already do the same?
Also it might make sense to do the visuals in shared so they are predicted, and only keep the hotspot in the server, but that is not a blocker

slarticodefast commented on Content.Shared/Lightning/Components/MatchboxComponent.cs:
```diff
@@ -0,0 +1,11 @@
+// TODO make changes in icons when different threshold reached
+// e.g. different icons for 10% 50% 100%
+[RegisterComponent, NetworkedComponent]
+public sealed partial class MatchboxComponent : Component
```
can you add documentation to the component?

slarticodefast commented on Content.Shared/Lightning/Components/MatchstickComponent.cs:
```diff
@@ -0,0 +1,35 @@
+    /// <summary>
+    ///     How long will matchstick last in seconds.
+    /// </summary>
+    [DataField("duration")]
```
```suggestion
    [DataField, AutoNetworkedField]
```

slarticodefast commented on Content.Shared/Lightning/Components/MatchstickComponent.cs:
```diff
@@ -0,0 +1,35 @@
+    ///     The time that the match will burn out. If null, that means the match is unlit.
+    /// </summary>
+    [AutoNetworkedField, AutoPausedField]
+    public TimeSpan? TimeMatchWillBurnOut = null;
```
needs time offset serializer

slarticodefast commented on Content.Shared/Lightning/Components/MatchstickComponent.cs:
```diff
@@ -0,0 +1,35 @@
+    /// <summary>
+    ///     Sound played when you ignite the matchstick.
+    /// </summary>
+    [DataField("igniteSound", required: true)] public SoundSpecifier IgniteSound = default!;
```
```suggestion
    [DataField]
    public SoundSpecifier? IgniteSound = null;
```
just make it nullable

slarticodefast commented on Content.Shared/Lightning/EntitySystems/MatchboxSystem.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Content.Shared.Light.EntitySystems;
+
+public sealed class SharedMatchboxSystem : EntitySystem
```
```suggestion
public abstract partial class SharedMatchboxSystem : EntitySystem
```

slarticodefast commented on Content.Shared/Lightning/EntitySystems/MatchboxSystem.cs:
```diff
@@ -0,0 +1,28 @@
+        SubscribeLocalEvent<MatchboxComponent, InteractUsingEvent>(OnInteractUsing, before: new[] { typeof(SharedStorageSystem) });
+    }
+
+    private void OnInteractUsing(EntityUid uid, MatchboxComponent component, InteractUsingEvent args)
```
```suggestion
    private void OnInteractUsing(Entity<MatchboxComponent> ent, ref InteractUsingEvent args)
```

slarticodefast commented on Content.Shared/Lightning/EntitySystems/MatchboxSystem.cs:
```diff
@@ -0,0 +1,28 @@
+        if (args.Handled || !EntityManager.TryGetComponent(args.Used, out MatchstickComponent? matchstick))
+            return;
+
+        if (matchstick.CurrentState == SmokableState.Unlit)
```
I would move this check into the Ignite function

slarticodefast commented on Content.Shared/Lightning/EntitySystems/SharedMatchstickSystem.cs:
```diff
@@ -0,0 +1,85 @@
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<MatchstickComponent, InteractUsingEvent>(OnInteractUsing);
```
```suggestion
        base.Initialize();

        SubscribeLocalEvent<MatchstickComponent, InteractUsingEvent>(OnInteractUsing);
```

slarticodefast commented on Content.Shared/Lightning/EntitySystems/SharedMatchstickSystem.cs:
```diff
@@ -0,0 +1,85 @@
+using Content.Shared.Light.Components;
+using Robust.Shared.Timing;
+
+namespace Content.Shared.Light.EntitySystems;
```
the smoking namespace delta was using in his PR probably makes more sense

slarticodefast commented on Content.Shared/Lightning/EntitySystems/SharedMatchstickSystem.cs:
```diff
@@ -0,0 +1,85 @@
+        args.Handled = true;
+    }
+
+    private void OnIsHotEvent(EntityUid uid, MatchstickComponent component, IsHotEvent args)
```
```suggestion
    private void OnIsHot(Entity<MatchstickComponent> ent, ref IsHotEvent args)
```

slarticodefast commented on Content.Shared/Lightning/EntitySystems/SharedMatchstickSystem.cs:
```diff
@@ -0,0 +1,85 @@
+
+    private void OnIsHotEvent(EntityUid uid, MatchstickComponent component, IsHotEvent args)
+    {
+        args.IsHot = component.CurrentState == SmokableState.Lit;
```
```suggestion
        args.IsHot |= component.CurrentState == SmokableState.Lit;
```
may overwrite another event subscription that already set it to true

slarticodefast commented on Content.Shared/Lightning/EntitySystems/SharedMatchstickSystem.cs:
```diff
@@ -0,0 +1,85 @@
+    public void Ignite(Entity<MatchstickComponent> matchstick, EntityUid user)
+    {
+        // Play Sound
+        _audio.PlayPredicted(matchstick.Comp.IgniteSound, matchstick, user, AudioParams.Default.WithVariation(0.125f).WithVolume(-0.125f));
```
the params should probably be moved into the sound specifier datafield

slarticodefast commented on Content.Shared/Lightning/EntitySystems/SharedMatchstickSystem.cs:
```diff
@@ -0,0 +1,85 @@
+
+        // Change state
+        SetState(matchstick, matchstick.Comp, SmokableState.Lit);
+        matchstick.Comp.TimeMatchWillBurnOut = _timing.CurTime + TimeSpan.FromSeconds(matchstick.Comp.Duration);
```
needs dirtying in case `Ignite` is called from the server

slarticodefast commented on Content.Shared/Lightning/EntitySystems/SharedMatchstickSystem.cs:
```diff
@@ -0,0 +1,85 @@
+
+    }
+
+    protected void SetState(EntityUid uid, MatchstickComponent component, SmokableState value)
```
same here

beck-thompson commented:
OK addressed the review! There were a few other minor cleanup things I did as well. I split it up into two commits, the first one is before the namespace switch so it might be easier to look at that.

slarticodefast commented on Content.Server/Matchstick/EntitySystems/MatchstickSystem.cs:
```diff
@@ -0,0 +1,24 @@
+    [Dependency] private readonly AtmosphereSystem _atmosphereSystem = default!;
+    [Dependency] private readonly TransformSystem _transformSystem = default!;
+
+    protected override void CreateMatchstickHotspot(Entity<MatchstickComponent> ent)
+    {
+        var xform = Transform(ent);
+
+        if (xform.GridUid is not { } gridUid)
+            return;
+
+        var position = _transformSystem.GetGridOrMapTilePosition(ent, xform);
+
+        _atmosphereSystem.HotspotExpose(gridUid, position, ent.Comp.BurnTemperature, 50, ent, true);
+    }
```
Ok, I had another look and I think we should just merge https://github.com/space-wizards/space-station-14/pull/36310 first and give matchsticks the `IgnitionSourceComponent` because this code is directly copy pasted from there.
That way we can completely kill the server side system here. All you need to do is use `SetIgnited` in the `IgnitionSourceSystem`.
I think the `IgnitionSource` namespace would also be work for matchsticks!
Sorry for missing this in the previous review and making you change this again

slarticodefast commented on Content.Shared/IgnitionSource/EntitySystems/SharedMatchstickSystem.cs:
```diff
@@ -0,0 +1,102 @@
+
+namespace Content.Shared.IgnitionSource.EntitySystems;
+
+public abstract class SharedMatchstickSystem : EntitySystem
```
```suggestion
public sealed partial class MatchstickSystem : EntitySystem
```
since they are empty we can just make this non-abstract and remove the server and client systems now

beck-thompson replied:
I actually left them like that on purpose so it would be easier for forks to make changes if they wanted (E.g they wouldn't have to make a new class and it would reduce merge conflicts) - we don't really seem to have a policy either way and I'm also unsure if any forks even would have any changes. Let me know what you think!

slarticodefast replied:
you can still do that since it is partial

beck-thompson replied:
> you can still do that since it is partial

From what I remember (I could be wrong!) it doesn't quite work because the shared class isn't abstract so when you make a child it gets all funky. It doesn't matter much either way though so I'll just remove em! It's usually done that way anyway

slarticodefast reviewed (changes requested):
One final thing, the rest looks good.

slarticodefast reviewed (approved):
Works great! Thanks for the PR!
